ui/src/utils/async_task_queue.rs

```diff
@@ -0,0 +1,128 @@
+use common::{
+    icons::outline::Shape as Icon,
+    state::pending_message::PendingMessage,
+    warp_runner::{ui_adapter::MessageEvent, WarpEvent},
+    WARP_EVENT_CH,
+};
+use dioxus_core::ScopeState;
+use dioxus_hooks::{use_ref, UseRef};
+use futures::{Future, StreamExt};
+use tokio::sync::{
+    mpsc::{UnboundedReceiver, UnboundedSender},
+    Mutex,
+};
+
+use once_cell::sync::Lazy;
+use std::sync::Arc;
+use uuid::Uuid;
+use warp::raygun::{AttachmentEventStream, AttachmentKind, Location};
+
+pub enum ListenerAction {
+    ToastAction {
+        title: String,
+        content: String,
+        icon: Option<Icon>,
+        timeout: u32,
+    },
+}
+
+pub struct ListenerChannel {
+    pub tx: UnboundedSender<ListenerAction>,
+    pub rx: Arc<Mutex<UnboundedReceiver<ListenerAction>>>,
+}
+
+// Channel for actions that should be done on the main thread
+pub static ACTION_LISTENER: Lazy<ListenerChannel> = Lazy::new(|| {
+    let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
+    ListenerChannel {
+        tx,
+        rx: Arc::new(Mutex::new(rx)),
+    }
+});
```

Comment:
I do wonder if this should, in some way, be localized instead of it being a static. Not required to change in this PR but just something to think about for the future since ideally it shouldnt be sitting as a static.

Reply:
i think it should. as this can be used as a base to expand in the future in case we need to do other offthread stuff that needs to send to the main thread